### ISSUE TITLE
CS GO Fails to update or install for Linux

### ISSUE THREAD

**Issue by KungFuJesus** [ValveSoftware/steam-for-linux#4660]
#### Your system information
- Steam client version: 1476379980
- Distribution (e.g. Ubuntu): Gentoo
- Opted into Steam client beta?: No
- Have you checked for system updates?: Yes

This happens for both the beta & non-beta versions of steam.  I also tried removing my entire steam library (rm'd the entire directory) and rm'd all traces of steam & preferences - followed by a reinstall.  It gets the entire game downloaded, then starts the install stage where it is likely copying/extracting files, and then it spits this on the console:

depotreconstruct.cpp (490) : Assertion Failed: pInfo->nNumWritesFinished > 0
Assert( Assertion Failed: pInfo->nNumWritesFinished > 0 ):depotreconstruct.cpp:490

Then it says "PAUSED: CORRUPT UPDATE FILES".  Every new launch of steam it tries to redo the installation procedure and fails.

I'm using a user-specified folder for a steam library that resides on a different file system, though I don't think that matters.  
1. Try to either update or newly install CS GO on Linux.

**Comment by KungFuJesus** [comment]
Just an added detail, I managed to install the game successfully by choosing a different library folder backed by an EXT4 file system.  The steam library folder it was failing to install to was an F2FS file system.

**Comment by tpruzina** [comment]
Same problem, nearly same setup (gentoo, F2FS).
Are you running current kernel (linuses tree)?

**Comment by tpruzina** [comment]
The process/thread that seems to be failing is "VerifyStagingFi", when I straced it it's spamming:

<pre>
...
semop(327686, {{0, 1, 0}}, 1)           = 0
semop(294917, {{0, -1, 0}}, 1)          = 0
semop(327686, {{0, 1, 0}}, 1)           = 0
semop(294917, {{0, -1, 0}}, 1)          = -1 EINTR (Interrupted system call)
--- SIGSTOP {si_signo=SIGSTOP, si_code=SI_USER, si_pid=20583, si_uid=1000} ---
--- stopped by SIGSTOP ---
....
semop(327686, {{0, 1, 0}}, 1)           = 0
semop(294917, {{0, -1, 0}}, 1)          = 0
semop(327686, {{0, 1, 0}}, 1)           = 0
semop(294917, {{0, -1, 0}}, 1)          = 0
semop(327686, {{0, 1, 0}}, 1)           = 0
madvise(0xdadff000, 1032192, MADV_DONTNEED) = 0
</pre>

Unfortunatedly that's not much helpful without REing how these mutexes work.
Anyways, if devs have any idea how to debug this or shed some light onto what's going on, I'm all ears.

**Comment by KungFuJesus** [comment]
I'm on 4.7.2 with the gentoo-sources ebuild, so I'm not entirely sure it's a regression on F2FS's end.  At least not a recent one.  Though until this point I had no issue.  I also am seeing unplayable framerates with L4D2, though that could be unrelated.  My console is spammed with "Creating process <somenum> for appid 550" where the numbers seem to increase forever.  Haven't tried it yet on an ext4 file system to see if it makes any difference for that issue.

**Comment by chinf** [comment]
I have the same problem, but I'm on Ubuntu 16.04 (kernel 4.4.0-43-generic) and ZFS (0.6.5.6) rather than F2FS. In my syslog I see:

gnome-session[3662]: ../tier1/fileio.cpp (3897) : Assertion Failed: CFileReader::Read must be called with a cubData value that is a multiple of the sector size when using unbuffered IO
gnome-session[3662]: depotreconstruct.cpp (490) : Assertion Failed: pInfo->nNumWritesFinished > 0
gnome-session[3662]: Assert( Assertion Failed: pInfo->nNumWritesFinished > 0 ):depotreconstruct.cpp:490
gnome-session[3662]: Installing breakpad exception handler for appid(steam)/version(1476379980)

Have not yet tried installing to an ext4 filesystem.

**Comment by tpgxyz** [comment]
Same here with F2FS file system 
[tpg@lazur ~]$ uname -a
Linux lazur 4.8.4-desktop-1omv #1 SMP PREEMPT Sun Oct 23 18:43:41 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux

**Comment by tpgxyz** [comment]
Looks like with latest Steam beta updates everything is fine under F2FS. Thanks for fixing it.

**Comment by lectrode** [comment]
Been having this issue for the past few months (same error messages and symptoms). I've tried everything from verifying files to fresh installing on a completely new computer and re-downloading all my games. The games experiencing the issue that I know of are CS:GO and Dota. Tested on both latest stable and latest beta versions of Steam for linux.

I'm using Btrfs for the main system partition. The games are stored on a larger hard drive formatted to NTFS. This setup was working fine until a few months ago (and still does for the rest of the games). Both hard drives are the standard mechanical ones, not SSD.

**I did find something that seems fairly interesting: it fails when the NTFS hard drive is connected via SATA. If I connect the same hard drive via USB as an external device (through a SATA to USB adapter), it works without issue.**

(There is nothing physically wrong with the SATA connections. As stated above I tried fresh installing everything on a new computer and had the same results)

(If you're wondering why I'm using NTFS instead of much better filesystems, its because this is usually part of a dual boot system and both the Windows games and Linux games are/will be stored on it)

Technical:
Kernel: 4.4.41-1-MANJARO
Steam: Built: Jan 19 2017 1:23:02; API: v017; package versions: 1484790260
